README.md
# good-clock
An **intuitive timer / clock** for **javascript** + **typescript**

## Install

    npm i good-clock

## Quick Links

 - [Getting Started](#getting-started)
 - [Clock Details](#clock-details)
 - [Actions](#actions)
 - [Setters](#setters)
 - [Getters](#getters)
 - [Helpers](#helpers)
 - [Examples](#examples)
 - [Defaults](#defaults)
 - [Improvements](#improvements)

## Getting Started

Setup

    import Clock from 'good-clock'       // for import method
    const Clock = require('good-clock')  // for require method

Create a clock object - *( or multiple if you want )*

    const clock = new Clock

Classic "setInterval" method

    clock.start(() => {
      console.log('tick-tock')
    }, 1000)

Dot Notation method

    clock.setCallback(() => { console.log('tick-tock') })
    clock.setTimeInterval(1000)
    clock.start()

Things get fun

    clock.start((value) => {
      console.log(value)
    }, clock.seconds(1), {
      {
        startValue: 10,
        endValue: 20,
        endCallback: () => {
          console.log("all done!")
        }
      }
    })

Everything works using both methods, this will do the same thing

    clock.setCallback((value) => { console.log(value) })
    clock.setTimeInterval(clock.seconds(1))
    clock.setStartValue(10)
    clock.setEndValue(20)
    clock.setEndCallback(() => { console.log("all done!") })
    clock.start()


## Clock Details

Each clock object contains one "clock" and one "wait". The clock acts like javascript's setInterval and the wait acts like javascript's setTimeout. Why use them? They have some awesome additions that make them more intuitive and quicker to implement. The clock has a built in value state which increments or decrements on every clock tick and is automagically passed to the clock's callback function. Unlike javascript's setInterval, you can change the clock's callback function entirely and the next tick will call the new callback.

The most important difference to realize is that good-clock is object oriented! You have to create a new clock object for every different clock you want to use. Calling clock.start(function) will NOT create a new clock, it only starts the clock its acting on. Each good-clock instance only has ONE wait as well! For example, **if you want two timers you DO NOT DO THIS**:

    // BAD TWO CLOCK IMPLEMENTATION
    // DO NOT DO THIS!!!
    
    const clock = new Clock
    
    clock.start(() => {
	    someFunction()
    })
    
    clock.start((value) => {
	    someOtherFunction(value)
    })

**Instead do this**:

    // GOOD TWO CLOCK IMPLEMENTATION
    // DO THIS
    
    const clockOne = new Clock
    const clockTwo = new Clock
    
    clockOne.start(() => {
	    someFunction()
    })
    
    clockTwo.start((value) => {
	    someOtherFunction(value)
    })



## Actions

| code | description | options |
|--|--|--|
| `clock.increment()` | Increments the clock's value by one valueInterval or by the optional argument | **number:** the amount to increment the clock's value. (optional) |
| `clock.decrement()` | Decrements the clock's value by one valueInterval or by the optional argument | **number:** the amount to decrement the clock's value. (optional) |
| `clock.start()` | Starts the clock, optionally sets it up. | **function:** see clock.setCallback() below. **number:** see clock.setTimeInterval() below. **ClockOptions:** see ClockOptions below. |
| `clock.stop()` | Stops the clock and resets it to startValue | none |
| `clock.stopAfterNextCallback()` | Stops the clock after the next clock tick and resets it to startValue. | none |
| `clock.pause()` | Pauses the clock. Does NOT change the clock's value and `clock.resume()` can then be used to restart the clock as if it had never stopped. | none |
| `clock.pauseAfterNextCallback()` | Pauses the clock after the next clock tick. Does NOT change the clock's value and `clock.resume()` can then be used to restart the clock as if it had never stopped. | none |
| `clock.resume()` | Restarts the clock without changing it's value. Remember that `clock.stop()` resets the value to the clock's startValue. So if you use `clock.stop()` to stop the clock and then resume to restart the clock, the clock WILL start at the clock's startValue. Use `clock.pause()` to stop a clock without reseting it's value to it's startValue. | none |
| `clock.wait()` | Calls a function after a determined wait time. | **function:** The function to be called after the wait time. **number:** The time to wait before calling the function. |
| `clock.cancelWaitWithCallback()` | Call this when the clock is waiting from `clock.wait()` to cancel the wait and immediately call its function. | none |
| `clock.cancelWaitWithoutCallback()` | Call this when the clock is waiting from `clock.wait()` to cancel the wait and NOT call its function. | none |

**ClockOptions**: type: object ( all properties are optional )
	

    {
	    valueInterval: number,
	    incrementing: boolean,
	    decrementing: boolean,
	    startValue: number,
	    startCallback: function,
	    endValue: number,
	    endCallback: function,
	    clampEndValue: boolean,
	    skipInitialCallback: boolean,
        incrementBeforeCallbacks: boolean,
	    incrementBeforeInitialCallback: boolean,
	    driftCorrectionOn: boolean
	}

For more details on **ClockOptions** see the corresponding setters below.
## Setters

| code | description |
|--|--|
| `clock.setCallback(function(value))` | Sets the callback of the clock. This is a function that's called on every clock tick. The function receives the clock's `value` as an argument. |
| `clock.setValue(number)` | Sets the value of the clock. This can be any number. The value will be incremented or decremented every clock tick based on the clock's valueInterval. |
| `clock.setTimeInterval(number)` | Sets the time interval of the clock in milliseconds. The clock is based on javascript's setTimeout function, so the time between clock callbacks is NOT perfectly precise. This time can vary quite a bit, sometimes 50ms or so. Each lag (or inaccurately triggered interval) adds on top of the last. So a clock with it's timeInterval set to 1000, (1 second) after 120 seconds (2 minutes) will be more like 120.5 or 121 seconds later. If you want the clock to sync with real world seconds / minutes / etc... see the `clock.setDriftCorrection()` method. |
| `clock.setValueInterval(number)` | Sets the amount to increase or decrease the clock's value on every clock tick. Whether it adds or subtracts is based on whether the clock is set to increment or decrement. |
| `clock.setIncrementing(boolean or undefined)` | If called without an argument or with `true` the clock will be set to increment. If called with `false` the clock will be set to decrement. |
| `clock.setDecrementing(boolean or undefined)` | If called without an argument or with `true` the clock will be set to decrement. If called with `false` the clock will be set to increment. |
| `clock.setStartValue(number)` | Sets the clock's start value. When calling `clock.start()` the clock will start at this value. If startValue is not set and `clock.start()` is called the clock will start with whatever clock.value is set as, and startValue will be automatically set to the clock.value. |
| `clock.setStartCallback(function)` | Sets a function that is called when `clock.start()` is called. |
| `clock.setEndValue(number)` | Set's the endValue of the clock. If the value is equal to this endValue, the clock will automatically stop. If `clock.isEndValueClamped` is `true` then the clock will also stop if it is greater than (when incrementing) or less than (when decrementing) this endValue. |
| `clock.setEndCallback(function)` | Sets a function that is called when the endValue causes the clock to stop. |
| `clock.setClampEndValue(number)` | Sets the value that the clock will automatically stop at (endValue). If `clock.isEndValueClamped` is `true` then the clock will also stop if it is greater than (when incrementing) or less than (when decrementing) the endValue. |
| `clock.setSkipInitialCallback(boolean)` | Sets whether or not the clock's first tick is when `clock.start()` is called, or one timeInterval after `clock.start()` is called. |
| `clock.setIncrementBeforeCallbacks(boolean)` | Sets whether or not the clock's value will increment / decrement before or after each callback. |
| `clock.setIncrementBeforeInitialCallback(boolean)` | Sets whether or not the clock's value will increment / decrement before the first clock tick. |
| `clock.setDriftCorrection(boolean)` | The clock is based on javascript's setTimeout function, so the time between clock callbacks is NOT perfectly precise. This time can vary quite a bit, sometimes 50ms or so. Each lag (or inaccurately triggered interval) adds on top of the last. So a clock with it's timeInterval set to 1000, (1 second) after 120 seconds (2 minutes) will be more like 120.5 or 121 seconds later. If you want the clock to sync with real world seconds / minutes / etc... set this to true. Note that the clock will still not fire at precisely the right time, this only ensures that the clock's error will not accumulate. If the clock's drift is greater than 50ms, driftCorrection is automatically disabled. If the clock's timeInterval is set to less than 50ms, driftCorrection is automatically disabled. |

## Getters

| code | description | type |
|--|--|--|
| `clock.value` | The value of the clock. | number |
| `clock.timeInterval` | The time interval of the clock. | number |
| `clock.valueInterval` | The value interval of the clock. The value interval is how much the value will change each clock tick. | number |
| `clock.isIncrementing` | `true` if the clock is set to increment, false if set to decrement. | boolean |
| `clock.isDecrementing` | `true` if the clock is set to decrement, `false` if set to increment. | boolean |
| `clock.startValue` | The value the clock will start at when ran with `clock.start()`. It is `undefined` until either directly set or the first time `clock.start()` is called. The first time `clock.start()` is called, it's startValue will be set to whatever the value was. (default = 1). Calling `clock.stop()` will set the clock's value back to this startValue. Calling `clock.start()` again, will start the clock at this startValue. | number |
| `clock.endValue` | The clock will automatically stop once the clock's value equals this value. | number |
| `clock.isEndValueClamped` | `true` if the clock is set to be clamped to the end value. `false` if not. (default = false) | boolean |
| `clock.isSkippingInitialCallback` | `true` if the clock is set to skip the initial callback when `clock.start()` is called. `false` if not. (default = false) | boolean |
| `clock.isIncrementingBeforeCallbacks` | `true` if the clock is set to increment / decrement before each callback instead of after. `false` if not. (default = false) | boolean |
| `clock.isIncrementingBeforeInitialCallback` | `true` if the clock is set to increment before the initial callback. `false` if not. | boolean |
| `clock.isDriftCorrectionOn` | `true` if drift correction is on. `false if not. (default = false) | boolean |
| `clock.isRunning` | `true` if the clock is running. `false` if not. | boolean |
| `clock.isPaused` | `true` if the clock is paused. `false` if not. | boolean |
| `clock.isStopped` | `true` if the clock is stopped. `false` if not. | boolean |
| `clock.isWaiting` | `true` if the clock's `clock.wait()` is waiting. `false` if not. | boolean |
| `clock.isAtStart` | `true` if the clock's value is equal to the clock's startValue. `false` if not. | boolean |
| `clock.isAtEnd` | `true` if the clock's value is equal to the clock's endValue. `false` if not. | boolean |
| `clock.willStopAfterNextCallback` | `true` if the clock is set to stop after the next callback after `clock.stop()` is called. `false` if not. | boolean |
| `clock.willPauseAfterNextCallback` | `true` if the clock is set to pause after the next callback after `clock.pause()` is called. `false` if not. | boolean |
| `

## Helpers

| code | description |
|--|--|
| `clock.seconds(number)` | Returns milliseconds. (input * 1000) |
| `clock.minutes(number)` | Returns milliseconds. (input * 60 * 1000) |
| `clock.hours(number)` | Returns milliseconds. (input * 60 * 60 * 1000) |

## Examples

Simple count to ten and stop, incrementing by one each second.

    clock.start((value) => {
	    console.log(value)
	}, clock.seconds(1), {
		endValue: 10,
	})

Count to ten, each one second, but pause at 5 for 3 seconds

    clock.start((value) => {
	    console.log(value)
	    if (value === 5) {
		    clock.pause()
		    clock.wait(() => {
			    clock.resume()
			}, clock.seconds(3))
		}
	}, clock.seconds(1), {
		endValue: 10,
	})

Easy decrementing

    clock.start((value) => {
	    console.log(value)
	}, clock.seconds(1), {
		decrementing: true,
		endValue: -10,
	})

Start at 15 and count to 20

    clock.start((value) => {
	    console.log(value)
	}, clock.seconds(1), {
		startValue: 15,
		endValue: 20,
	})

Pass the endValue, but since it's clamped it stops at the endValue

    clock.start((value) => {
	    console.log(value)
	}, clock.seconds(1), {
		valueInterval: 3,
		endValue: 10,
		clampEndValue: true,
	})

Change the value interval mid-flight

    clock.start((value) => {
	    clock.setValueInterval(value + 1)
	    console.log(value)
	}, clock.seconds(1), {
		endValue: 20,
	})

Change the callback function mid-flight ( after 5 second wait )

    clock.start((value) => {
		console.log(value)
	}, clock.seconds(1), {
		endValue: 10,
	})
	
	clock.wait(() => {
		clock.setCallback((value) => {
			console.log("value: " + value)
		})
	}, clock.seconds(5))

## Defaults

    callback = null
    value = 0
    timeInterval = 1000
    valueInterval = 1
    incrementing = true
    decrementing = false
    startValue = undefined
    startCallback = null
    endValue = undefined
    endCallback = null
    clampEndValue = false
    skipInitialCallback = false
    incrementBeforeCallbacks = false
    incrementBeforeInitialCallback = false
    isRunning = false
    isPaused = false
    isStopped = true
    waitCallback = null
    isWaiting = false
    willStopAfterNextCallback = false
    willPauseAfterNextCallback = false


## Improvements

If you would like to contribute improvements please [create issues](https://github.com/dtpietrzak/good-clock/issues) or [create pull requests](https://github.com/dtpietrzak/good-clock/pulls) on the github page. 

CHANGELOG
1.2.0 -- 2020-01-10
* Added incrementBeforeCallbacks
* Added more tests
* Updated Readme

1.1.4 -- 2020-01-09
* Updated Readme

1.1.2 -- 2020-01-09
* Updated Readme

1.1.1 -- 2020-01-07
* Changed console.logs to console.warns when methods are called with the wrong type
* Updated Readme

1.1.0 -- 2020-01-04
* Added optional drift correction and exports.module syntax

1.0.0 -- 2020-01-03
* Initial release

test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const clock_1 = __importDefault(require("./clock"));
const clock = new clock_1.default;
console.log("test begins");
clock.setValue(-5);
clock.start((value) => {
    console.log(value);
    console.log(new Date());
}, 100, {
    endValue: 20,
    endCallback: () => {
        console.log('test ends');
    },
    valueInterval: (last_value, last_interval) => {
        console.log(last_interval);
        return (1);
    }
});


test.ts
import Clock from './clock'

const clock = new Clock;

console.log("test begins")

clock.setValue(-5)

clock.start((value) => {
  console.log(value)
  console.log(new Date())
}, 100, {
  endValue: 20,
  endCallback: () => {
    console.log('test ends')
  },
  valueInterval: (last_value, last_interval) => {
    console.log(last_interval)
    return (1)
  }
})


src/index.ts
export type Value = number
export type Milliseconds = number
export type ClockTickCallback = ((value: Value) => void | boolean) | null

export class Clock {
  private _setinterval: ReturnType<typeof setInterval> | null
  private _settimeout: ReturnType<typeof setTimeout> | null
  private _callback: ClockTickCallback
  private _value: Value
  private _timeInterval: Milliseconds
  private _valueInterval: Value
  private _incrementing: boolean
  private _decrementing: boolean
  private _startValue: Value | undefined
  private _startCallback: ClockTickCallback
  private _endValue: Value | undefined
  private _endCallback: ClockTickCallback
  private _clampEndValue: boolean
  private _skipInitialCallback: boolean
  private _incrementBeforeCallbacks: boolean
  private _incrementBeforeInitialCallback: boolean
  private _isRunning: boolean
  private _isPaused: boolean
  private _isStopped: boolean
  private _waitCallback: ClockTickCallback
  private _isWaiting: boolean
  private _willStopAfterNextCallback: boolean
  private _willPauseAfterNextCallback: boolean

  // drift correction stuff
  private _epochStarted: number
  private _expected: number
  private _drifted: number
  private _hadDriftError: boolean
  private _driftCorrectionOn: boolean

  constructor() {
    this._setinterval = null
    this._settimeout = null
    this._callback = null
    this._value = 0
    this._timeInterval = 1000
    this._valueInterval = 1
    this._incrementing = true
    this._decrementing = false
    this._startValue = undefined
    this._startCallback = null
    this._endValue = undefined
    this._endCallback = null
    this._clampEndValue = false
    this._skipInitialCallback = false
    this._incrementBeforeCallbacks = false
    this._incrementBeforeInitialCallback = false
    this._isRunning = false
    this._isPaused = false
    this._isStopped = true
    this._waitCallback = null
    this._isWaiting = false
    this._willStopAfterNextCallback = false
    this._willPauseAfterNextCallback = false

    // drift correction stuff
    this._epochStarted = 0
    this._expected = 0
    this._drifted = 0
    this._hadDriftError = false
    this._driftCorrectionOn = false
  }


  // private stuff  -  private stuff  -  private stuff
  // private stuff  -  private stuff  -  private stuff
  // private stuff  -  private stuff  -  private stuff

  private _incrementOrDecrement = () => {
    // increment if only increment option is set to true
    // decrement if decrement or both are set to true
    // do nothing if both are false

    // this isn't intuitive until you go to use the function
    if (this._incrementing && !(this._decrementing)) {
      this.increment()
    } else if (this._decrementing) {
      this.decrement()
    }
  }

  private _runClockCallback = () => {
    if (typeof this._callback === 'function') {

      if (this._incrementBeforeCallbacks === true) {
        // check if should, and if should then inc or dec
        this._incrementOrDecrement()
      }

      if (this._endValue) {
        if (this._decrementing) {
          if (this._value <= this._endValue) {
            this.stop()
            if (this._clampEndValue) this._value = this._endValue
            if (this._endCallback) this._endCallback(this._value)
          }
        } else {
          if (this._value >= this._endValue) {
            this.stop()
            if (this._clampEndValue) this._value = this._endValue
            if (this._endCallback) this._endCallback(this._value)
          }
        }
      }

      // stop if set it to end after next callback
      if (this._willStopAfterNextCallback) {
        this.stop()
        this._willStopAfterNextCallback = false
      }

      // pause if set it to end after next callback
      if (this._willPauseAfterNextCallback) {
        this.pause()
        this._willPauseAfterNextCallback = false
      }

      // this if statement is for stopping the clock if the callback returns true
      if (this._callback(this._value)) this.stop()
    }

    if (this._incrementBeforeCallbacks === false) {
      // check if should, and if should then inc or dec
      this._incrementOrDecrement()
    }
  }

  // these intervals are off. last_interval keeps sending as a function
  private _calcTimeInterval = (interval: Milliseconds): Milliseconds => {
    if (typeof interval === 'number') {
      return interval
    } else {
      return this._calcTimeInterval(this._timeInterval)
    }
  }

  // these intervals are off. last_interval keeps sending as a function
  private _calcValueInterval = (interval: Milliseconds): Value | Value => {
    if (typeof interval === 'number') {
      return interval
    } else {
      return this._calcValueInterval(this._valueInterval)
    }
  }

  private _runClock = () => {
    this._isRunning = true
    this._isPaused = false
    this._isStopped = false

    this._setinterval = setTimeout(() => {
      this._runClock()
      // if callback is set, run it
      this._runClockCallback()
    }, this._calcTimeInterval(this._timeInterval) - this._driftCorrection())
  }

  // drift correction
  private _driftCorrection = () => {
    if (!(this._hadDriftError) && this._driftCorrectionOn) { // this is how we ignore drift correction
      if (this._expected !== 0) {
        // handles all ticks after the first
        // we know what the drift is for the last tick
        // so we'll simply subtract the last drift value
        this._drifted = (Date.now() - this._epochStarted) - this._expected
        console.log('drift', this._drifted)
        this._expected = this._expected + this._timeInterval
        console.log('expected', this._expected)
        console.log('actual', Date.now() - this._epochStarted + this._timeInterval)

        // if the drift messed up
        // we'll set all future ticks to ignore drift correction
        if (
          this._drifted < 0 ||
          this._drifted > this._timeInterval ||
          this._drifted > 50
        ) {
          this._hadDriftError = true
          return 0
        }
        return this._drifted
      } else {
        // handles the first tick
        // we dont know what the drift is yet,
        this._epochStarted = Date.now()
        this._expected = this._timeInterval
        return 0
      }
    }
    return 0
  }

  // public getters  -  public getters  -  public getters
  // public getters  -  public getters  -  public getters
  // public getters  -  public getters  -  public getters

  get value(): Value { return (this._value) }
  get timeInterval(): Milliseconds { return (this._timeInterval) }
  get valueInterval(): Value { return (this._valueInterval) }
  get isIncrementing(): boolean {
    if (this._decrementing) {
      return (false)
    } else if (this._incrementing) {
      return (true)
    } else {
      return (false)
    }
  }
  get isDecrementing(): boolean { return (this._decrementing) }
  get startValue(): Value | undefined { return (this._startValue) }
  get endValue(): Value | undefined { return (this._endValue) }
  get isEndValueClamped(): boolean { return this._clampEndValue }
  get isSkippingInitialCallback(): boolean { return this._skipInitialCallback }
  get isIncrementingBeforeCallbacks(): boolean { return this._incrementBeforeCallbacks }
  get isIncrementingBeforeInitialCallback(): boolean { return this._incrementBeforeInitialCallback }
  get isDriftCorrectionOn(): boolean { return this._driftCorrectionOn }
  get isRunning(): boolean { return (this._isRunning) }
  get isPaused(): boolean { return (this._isPaused) }
  get isStopped(): boolean { return (this._isStopped) }
  get isWaiting(): boolean { return (this._isWaiting) }
  get isAtStart(): boolean { return (this._value === this._startValue) }
  get isAtEnd(): boolean { return (this._value === this._endValue) }
  get willStopAfterNextCallback(): boolean { return (this._willStopAfterNextCallback) }
  get willPauseAfterNextCallback(): boolean { return (this._willPauseAfterNextCallback) }


  // public setters  -  public setters  -  public setters
  // public setters  -  public setters  -  public setters
  // public setters  -  public setters  -  public setters

  // callback
  public setCallback = (callback: ClockTickCallback) => {
    if (typeof callback !== 'function') {
      console.warn("Clock Error: Clock.setCallback() input must be a function")
    } else {
      this._callback = callback
    }
  }

  // value
  public setValue = (value: Value) => {
    if (typeof value !== 'number') {
      console.warn("Clock Error: Clock.setValue() input must be a number")
    } else {
      this._value = value
    }
  }

  // time interval
  public setTimeInterval = (newValue: Milliseconds) => {
    if (typeof newValue !== 'number') {
      console.warn("Clock Error: Clock.setTimeInterval() input must be a number")
    } else {
      this._timeInterval = newValue
      if (newValue < 99) this._driftCorrectionOn = false
    }
  }

  // value interval
  public setValueInterval = (newValue: Value) => {
    if (typeof newValue !== 'number') {
      console.warn("Clock Error: Clock.setValueInterval() input must be a number")
    } else {
      this._valueInterval = newValue
    }
  }

  // set it to incrementing state,
  // if an option isnt passed, it sets it to increment
  public setIncrementing = (incrementing?: boolean) => {
    if (incrementing === undefined) {
      this._incrementing = true
      this._decrementing = false
    } else {
      if (typeof incrementing === 'boolean') {
        this._incrementing = incrementing
        this._decrementing = !incrementing
      } else {
        console.warn("Clock Error: Clock.setIncrementing() input must be a boolean or undefined")
      }
    }
  }

  // set it to decrementing state,
  // if an option isnt passed, it sets it to decrement
  public setDecrementing = (decrementing?: boolean) => {
    if (decrementing === undefined) {
      this._incrementing = false
      this._decrementing = true
    } else {
      if (typeof decrementing === 'boolean') {
        this._incrementing = !decrementing
        this._decrementing = decrementing
      } else {
        console.warn("Clock Error: Clock.setDecrementing() input must be a boolean or undefined")
      }
    }
  }

  // start value
  public setStartValue = (startValue: Value) => {
    if (typeof startValue !== 'number') {
      console.warn("Clock Error: Clock.setStartValue() input must be a number")
    } else {
      this._startValue = startValue
    }
  }

  public setStartCallback = (startCallback: ClockTickCallback) => {
    if (typeof startCallback !== 'function') {
      console.warn("Clock Error: Clock.setStartCallback() input must be a function")
    } else {
      this._startCallback = startCallback
    }
  }

  // end value
  public setEndValue = (endValue: Value) => {
    if (typeof endValue !== 'number') {
      console.warn("Clock Error: Clock.setEndValue() input must be a number")
    } else {
      this._endValue = endValue
    }
  }

  public setEndCallback = (endCallback: ClockTickCallback) => {
    if (typeof endCallback !== 'function') {
      console.warn("Clock Error: Clock.setEndCallback() input must be a function")
    } else {
      this._endCallback = endCallback
    }
  }

  public setClampEndValue = (isClamped: boolean) => {
    if (typeof isClamped !== 'boolean') {
      console.warn("Clock Error: Clock.setClampEndValue() input must be a boolean")
    } else {
      this._clampEndValue = isClamped
    }
  }

  public setSkipInitialCallback = (skipInitialCallback: boolean) => {
    if (typeof skipInitialCallback !== 'boolean') {
      console.warn("Clock Error: Clock.setSkipInitialCallback() input must be a boolean")
    } else {
      this._skipInitialCallback = skipInitialCallback
    }
  }

  public setIncrementBeforeCallbacks = (incrementBeforeCallbacks: boolean) => {
    if (typeof incrementBeforeCallbacks !== 'boolean') {
      console.warn("Clock Error: Clock.setIncrementBeforeCallbacks() input must be a boolean")
    } else {
      this._incrementBeforeCallbacks = incrementBeforeCallbacks
    }
  }

  public setIncrementBeforeInitialCallback = (incrementBeforeInitialCallback: boolean) => {
    if (typeof incrementBeforeInitialCallback !== 'boolean') {
      console.warn("Clock Error: Clock.setIncrementBeforeInitialCallback() input must be a boolean")
    } else {
      this._incrementBeforeInitialCallback = incrementBeforeInitialCallback
    }
  }

  public setDriftCorrection = (driftCorrection: boolean) => {
    if (typeof driftCorrection !== 'boolean') {
      console.warn("Clock Error: Clock.setDriftCorrection() input must be a boolean")
    } else {
      this._driftCorrectionOn = driftCorrection
    }
  }

  // public actions  -  public actions  -  public actions
  // public actions  -  public actions  -  public actions
  // public actions  -  public actions  -  public actions

  // increment
  public increment = (interval?: Value): Value | boolean => {
    if (interval) {
      if ((typeof interval !== 'number')) {
        console.warn("Clock Error: Clock.increment() input must be a number")
        return (false)
      }
      this._value = this._value + this._calcValueInterval(interval)
      return (this._value)
    } else {
      this._value = this._value + this._calcValueInterval(this._valueInterval)
      return (this._value)
    }
  }

  // decrement
  public decrement = (interval?: Value): Value | boolean => {
    if (interval) {
      if ((typeof interval !== 'number')) {
        console.warn("Clock Error: Clock.decrement() input must be a number")
        return (false)
      }
      this._value = this._value - this._calcValueInterval(interval)
      return (this._value)
    } else {
      this._value = this._value - this._calcValueInterval(this._valueInterval)
      return (this._value)
    }
  }

  // this starts a repeating clock that calls the clock's callback or an optional callback argument passed to the function. if the internal callback and the optional callback arent set, the clock still increments, but nothing happens. if both are set, only the optional argument callback is called
  public start = (
    callback?: ClockTickCallback,
    timeInterval?: Milliseconds,
    options?: {
      valueInterval?: Value,
      incrementing?: boolean,
      decrementing?: boolean,
      startValue?: Value,
      startCallback?: ClockTickCallback,
      endValue?: Value,
      endCallback?: ClockTickCallback,
      clampEndValue?: boolean,
      skipInitialCallback?: boolean,
      incrementBeforeCallbacks?: boolean,
      incrementBeforeInitialCallback?: boolean,
      driftCorrectionOn?: boolean,
    }
  ) => {
    if (this._setinterval == null) {
      if (callback) this._callback = callback
      if (timeInterval) this._timeInterval = timeInterval
      if (options?.incrementing) this._incrementing = options.incrementing
      if (options?.decrementing) this._decrementing = options.decrementing
      if (options?.skipInitialCallback)
        this._skipInitialCallback = options.skipInitialCallback
      if (options?.incrementBeforeCallbacks)
        this._incrementBeforeCallbacks = options.incrementBeforeCallbacks
      if (options?.incrementBeforeInitialCallback)
        this._incrementBeforeInitialCallback = options.incrementBeforeInitialCallback
      if (options?.valueInterval) this._valueInterval = options.valueInterval
      if (options?.startValue) this._startValue = options.startValue
      if (options?.startCallback) this._startCallback = options.startCallback
      if (options?.endValue) this._endValue = options.endValue
      if (options?.clampEndValue) this._clampEndValue = options.clampEndValue
      if (options?.endCallback) this._endCallback = options.endCallback
      if (options?.driftCorrectionOn) this._driftCorrectionOn = options.driftCorrectionOn

      if (this._timeInterval < 99) this._driftCorrectionOn = false

      if (this._startValue !== undefined) {
        this._value = this._startValue
      } else {
        this._startValue = this._value
      }

      // if the optional incrementOnInitialCallback is set, do it
      if (this._incrementBeforeInitialCallback) this._incrementOrDecrement()

      // if start callback is set, run it
      if (this._startCallback) this._startCallback(this._value)

      // if callback is set and it's not set to initially be skipped
      if (!(this._skipInitialCallback)) this._runClockCallback()

      // reset expected epoch and drift
      this._drifted = 0
      this._expected = 0

      // start the clock
      this._runClock()
    }
  }

  // stop, just like a CD player. it stops playback and restarts the clock
  public stop = () => {
    this._isRunning = false
    this._isPaused = false
    this._isStopped = true

    if (this._setinterval != null) {
      clearInterval(this._setinterval)
      this._setinterval = null
    }
    if (this._startValue) this._value = this._startValue
  }
  public stopAfterNextCallback = () => {
    this._willStopAfterNextCallback = true
  }

  // pause, just like a CD player. it stops playback, but doesn't restart the clock
  public pause = () => {
    this._isRunning = false
    this._isPaused = true
    this._isStopped = false

    if (this._setinterval != null) {
      clearInterval(this._setinterval)
      this._setinterval = null
    }
  }
  public pauseAfterNextCallback = () => {
    this._willPauseAfterNextCallback = true
  }

  // resume, just like a CD player. resumes playback after it was paused
  public resume = () => {
    if (this._setinterval == null) {
      // reset expected epoch and drift
      this._drifted = 0
      this._expected = 0
      this._runClock()
    }
  }

  // wait functions
  public wait = (callback: ClockTickCallback, waitTime: number) => {
    this._isWaiting = true

    this._waitCallback = callback

    this._settimeout = setTimeout(() => {
      this._isWaiting = false

      if (callback) {
        this._waitCallback = null
        callback(this._value)
      }
    }, waitTime)
  }

  // cancel wait with callback
  public cancelWaitWithCallback = () => {
    if (this._settimeout != null) {
      clearTimeout(this._settimeout)
      this._settimeout = null
    }
    if (this._waitCallback) {
      this._waitCallback(this._value)
    }
    this._waitCallback = null
  }

  // cancel wait with callback
  public cancelWaitWithoutCallback = () => {
    if (this._settimeout != null) {
      clearTimeout(this._settimeout)
      this._settimeout = null
      this._waitCallback = null
    }
  }

  // public converters  -  public converters  -  public converters
  // public converters  -  public converters  -  public converters
  // public converters  -  public converters  -  public converters

  // easy time conversions
  public seconds = (seconds: number): number => (seconds * 1000)
  public minutes = (minutes: number): number => (minutes * 60 * 1000)
  public hours = (hours: number): number => (hours * 60 * 60 * 1000)
}

module.exports = Clock
export default Clock